vivisect/analysis/generic/symswitchcase.py

```diff
@@ -0,0 +1,1307 @@
+
+            # peel off o_subs and size-limiting o_ands and o_sextends
+            if isinstance(symobj, o_and) and symobj.kids[1].isDiscrete() and symobj.kids[1].solve() in e_bits.u_maxes:
+                mask = symobj.kids[1].solve()
+                pass    # this wrapper is a size-limiting bitmask
+
+            elif isinstance(symobj, o_sub):
+                offset += symobj.kids[1].solve() # this is an offset, used to rebase the index into a different pointer array
+
+            elif isinstance(symobj, o_add):
+                offset -= symobj.kids[1].solve() # this is an offset, used to rebase the index into a different pointer array
+
+            elif isinstance(symobj, o_sextend):
+                pass # sign-extension is irrelevant for indexes
+
+            # anything else and we're done peeling
+            else:
+                break
+
+        logger.debug(" longpath DONE: (%r) %r" % (offset, symobj))
+        # now figure what was cut, and what impact it has.
+        self.baseIdx = symobj
+        self.baseoff = offset
+
+        return symobj, offset
+        
+    def getNormalizedConstraints(self):
+        '''
+        takes in a complex index symbolik state
+        returns a list of tuples of (SYMT_CON_??, offset)
+        '''
+        retcons = []
+        baseIdx, baseoff = self.getBaseSymIdx()
+
+        for con in self.getBoundingCons(baseIdx): # merge the two
+            cons = con.cons
+
+            if not cons.symtype in (SYMT_CON_GT, SYMT_CON_GE, SYMT_CON_LT, SYMT_CON_LE): 
+                #logger.debug("SKIPPING NON-LIMITING: cons = %s", repr(cons))
+                #return None, None, None
+                continue
+            
+            if cons.kids[1].symtype == SYMT_CONST:
+                symvar = cons.kids[0]
+                symcmp = cons.kids[1]
+            
+            elif cons.kids[0].symtype == SYMT_CONST:
+                symcmp = cons.kids[0]
+                symvar = cons.kids[1]
+            
+            else:
+                # neither side of the constraint is a CONST.  this constraint does 
+                # not set static bounds on idx
+                #logger.debug("SKIPPING (non-discrete): cons = %s", repr(cons))
+                continue
+
+            if not contains(symvar, baseIdx):
+                logger.debug("Constraint not based on Index: %r" % cons)
+                continue
+
+            if cons.symtype == SYMT_CON_LT and symcmp.solve() == 0:
+                #logger.debug("SKIPPING Constraint Checking for Zero: %r" % cons)
+                continue
+
+            conthing, consoff = peelIdxOffset(symvar)
+
+            if conthing != baseIdx:
+                logging.info("FAIL: %r  != %r" % (conthing, baseIdx))
+                #raw_input("FAIL: %r  != %r" % (conthing, baseIdx))
+                continue
+
+            logger.debug("GOOD: %r\n\t%r\n\t%r\t%r + %r" % (cons, symvar, conthing, consoff, symcmp))
```

Comment:
I'm gonna have no idea what this means when I see this in any logs. Same for the "FAIL" message a few lines up.

Reply:
yeah, i can clean up these log messages.

Reply:
done